# Logistic Regression Project

**Problem -**

we have given a portugal bank data and we have to find something useful from that data. our task is to predict wheather the customer of that portuguese bank are going to opt for the term deposit or not

# Section 1- Data Loading

In [46]:
#import the required libraries
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [47]:
#reading csv data
data= pd.read_csv('/content/portu_banking_final.csv')
data.head(78)

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (0) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


age          job  marital  ... euribor3m nr_employed  y
0      44  blue-collar  married  ...     4.963      5228.1  0
1      53   technician  married  ...     4.021      5195.8  0
2      28   management   single  ...     0.729      4991.6  1
3      39     services  married  ...     1.405      5099.1  0
4      55      retired  married  ...     0.869      5076.2  1
..    ...          ...      ...  ...       ...         ... ..
73     45     services  married  ...     4.857      5191.0  0
74  35-30   management  married  ...     4.960      5228.1  0
75     45     services  married  ...     4.962      5228.1  0
76     43   technician        D  ...     4.858      5191.0  0
77     54  blue-collar  married  ...     4.857      5191.0  0

[78 rows x 21 columns]

With the above data our task is to determine wheather a consumer will take the term deposit or not. If person is taking term deposit then it the output "y" will be 1 else it will be 0

In [48]:
data.shape

(41188, 21)

In [49]:
data.describe()

duration      campaign  ...   nr_employed             y
count  41188.000000  41188.000000  ...  41188.000000  41188.000000
mean     258.285010      2.567593  ...   5167.035911      0.112654
std      259.279249      2.770014  ...     72.251528      0.316173
min        0.000000      1.000000  ...   4963.600000      0.000000
25%      102.000000      1.000000  ...   5099.100000      0.000000
50%      180.000000      2.000000  ...   5191.000000      0.000000
75%      319.000000      3.000000  ...   5228.100000      0.000000
max     4918.000000     56.000000  ...   5228.100000      1.000000

[8 rows x 10 columns]

In [50]:
data['y'].value_counts()

0    36548
1     4640
Name: y, dtype: int64

it is clear that very few people, out of 41188 only 4640 consumer  took the term deposit and majority of people didnt opt for term deposit.

In [51]:
ones_count= int(4640)
zeroes_count= int(36548)
ratio_oz = ones_count/zeroes_count
print('The ratio of', ones_count, 'and', zeroes_count,'is', ratio_oz )




The ratio of 4640 and 36548 is 0.126956331399803


In [52]:
#it is cleared that ration of ones and zeroes are not balanced

## section 2-Data Cleaning

to clean the data we nedd to do some basic analysis such as use of dtypes, describe and info functions

Here we can see that age has a datatype "object" but camparing with the UCL data the dtype should be numeric. lets see whats wrong with the age column 

In [53]:
data.dtypes

age                object
job                object
marital            object
education          object
default            object
housing            object
loan               object
contact            object
month              object
day_of_week        object
duration            int64
campaign            int64
pdays             float64
previous            int64
poutcome           object
emp_var_rate      float64
cons_price_idx    float64
cons_conf_idx     float64
euribor3m         float64
nr_employed       float64
y                   int64
dtype: object

age has datatype 'object' that needs to be changed into integer for better understanding of data



In [54]:
data['age'].unique()

array(['44', '53', '28', '39', '55', '30', '37', '36', '27', '34', '41',
       '33', '26', '52', '35', '40', '32', '49', '38', '47', '46', '29',
       '54', '42', '72', '48', '43', '56', '31', '24', '68', '59', '50',
       '45', '35-30', '25', '57', '63', '58', '60', '30-35', '64', '51',
       '23', '20', '74', '80', '61', '25-30', '62', '75', '21', '82',
       '77', '18-25', '30-32', '70', '50-55', '76', '73', '66', '22',
       '71', '19', '79', '88', '65', '67', '81', '18', '84', '69', '98',
       '85', '83', '78', '92', '45-50', '20-25', '20-40', '35-40', '86',
       '94', '17', '91', '89', '87', '95', '30-40', 56, 39, 30, 53, 37,
       22, 45, 31, 24, 50, 58, 33, 27, 28, 32, 51, 41, 49, 43, 44, 35, 29,
       26, 57, 25, 42, 48, 46, 67, 38, 34, 54, 40, 23, 47, 62, 75, 55, 36,
       52, 59, 78, 70, 60, 20, 66, 61, 21, 88, 18, 68, 71, 73, 80, 79, 63,
       76, 83, 65, 19, 86, 17, 72, 69, 81, 64, 77, 74, 85, 82],
      dtype=object)

reason why data has datatype "object" is becauses age range is also included in age column .To make things better we will replace the age range with mean value of that age range. for example if age range is 20-40, we can replac that value with mean of that age range which is 30

In [55]:
# check the osition of all the age range value 
import re
s=""
i=-1
for ag in data.age:
  i+=1
  a=re.search("\d\d-\d\d", str(ag))
  if a!=None:
    print(a)
    print(i)

<re.Match object; span=(0, 5), match='35-30'>
74
<re.Match object; span=(0, 5), match='30-35'>
149
<re.Match object; span=(0, 5), match='25-30'>
480
<re.Match object; span=(0, 5), match='18-25'>
858
<re.Match object; span=(0, 5), match='30-32'>
967
<re.Match object; span=(0, 5), match='50-55'>
1085
<re.Match object; span=(0, 5), match='30-35'>
1457
<re.Match object; span=(0, 5), match='30-35'>
2571
<re.Match object; span=(0, 5), match='45-50'>
6475
<re.Match object; span=(0, 5), match='20-25'>
6987
<re.Match object; span=(0, 5), match='20-40'>
7190
<re.Match object; span=(0, 5), match='35-40'>
9832
<re.Match object; span=(0, 5), match='30-40'>
28550


In [56]:
data.loc[[74]]

age         job  marital  ... euribor3m nr_employed  y
74  35-30  management  married  ...      4.96      5228.1  0

[1 rows x 21 columns]

In [57]:
# replacing all the age range with theire respective mean value
data['age']=data['age'].replace(['35-30'],33)
data['age']=data['age'].replace(['30-35'],33)
data['age']=data['age'].replace(['25-30'],28)
data['age']=data['age'].replace(['18-25'],22)
data['age']=data['age'].replace(['30-32'],31)
data['age']=data['age'].replace(['50-55'],53)
data['age']=data['age'].replace(['50-55'],53)
data['age']=data['age'].replace(['45-50'],48)
data['age']=data['age'].replace(['20-25'],23)
data['age']=data['age'].replace(['20-40'],30)
data['age']=data['age'].replace(['35-40'],38)
data['age']=data['age'].replace(['30-40'],33)


In [58]:
data.loc[[9832]]

age          job  marital  ... euribor3m nr_employed  y
9832  38  blue-collar  married  ...      4.96      5228.1  0

[1 rows x 21 columns]

we can see now that the age range is now converted into the mean value and now we can change the datatype of age column

In [59]:
data['age'].unique()

array(['44', '53', '28', '39', '55', '30', '37', '36', '27', '34', '41',
       '33', '26', '52', '35', '40', '32', '49', '38', '47', '46', '29',
       '54', '42', '72', '48', '43', '56', '31', '24', '68', '59', '50',
       '45', 33, '25', '57', '63', '58', '60', '64', '51', '23', '20',
       '74', '80', '61', 28, '62', '75', '21', '82', '77', 22, 31, '70',
       53, '76', '73', '66', '22', '71', '19', '79', '88', '65', '67',
       '81', '18', '84', '69', '98', '85', '83', '78', '92', 48, 23, 30,
       38, '86', '94', '17', '91', '89', '87', '95', 56, 39, 37, 45, 24,
       50, 58, 27, 32, 51, 41, 49, 43, 44, 35, 29, 26, 57, 25, 42, 46, 67,
       34, 54, 40, 47, 62, 75, 55, 36, 52, 59, 78, 70, 60, 20, 66, 61, 21,
       88, 18, 68, 71, 73, 80, 79, 63, 76, 83, 65, 19, 86, 17, 72, 69, 81,
       64, 77, 74, 85, 82], dtype=object)

In [60]:
#change data type of age 
data['age']=data['age'].apply(pd.to_numeric)
print(data.age.dtypes)

int64


now lets find out errors like these in all the column.

in almost all the column consisting of categorical data has values like "unknown" which is nothing but NaN value. we will replace these values with the other value which ahs the highest number of mode

In [61]:
data['marital'].value_counts()

married     24923
single      11565
divorced     4608
unknown        80
M               5
D               4
S               3
Name: marital, dtype: int64

In [62]:
data['marital']=data["marital"].replace("unknown","married")
data['marital']=data["marital"].replace("M","married")
data['marital']=data["marital"].replace("S","single")
data['marital']=data["marital"].replace("D","divorced")
data.marital.value_counts()

married     25008
single      11568
divorced     4612
Name: marital, dtype: int64

In [63]:
data.default.value_counts()

no         32588
unknown     8597
yes            3
Name: default, dtype: int64

In [64]:
data['default']=data['default'].replace('unknown','no')
data.default.value_counts()

no     41185
yes        3
Name: default, dtype: int64

In [65]:
data.education.value_counts()

university.degree      12168
high.school             9515
basic.9y                6035
professional.course     5243
basic.4y                4164
basic.6y                2286
unknown                 1731
illiterate                18
basic-4y                  12
basic-9y                  10
basic-6y                   6
Name: education, dtype: int64

In [66]:
data['education']=data['education'].replace('unknown','university.degree')
data.education.value_counts()

university.degree      13899
high.school             9515
basic.9y                6035
professional.course     5243
basic.4y                4164
basic.6y                2286
illiterate                18
basic-4y                  12
basic-9y                  10
basic-6y                   6
Name: education, dtype: int64

In [67]:
data.housing.value_counts()

yes        21576
no         18622
unknown      990
Name: housing, dtype: int64

In [68]:
data['housing']=data['housing'].replace('unknown','yes')
data.housing.value_counts()

yes    22566
no     18622
Name: housing, dtype: int64

In [69]:
data.loan.value_counts()

no         33950
yes         6248
unknown      990
Name: loan, dtype: int64

In [70]:
data['loan']=data['loan'].replace('unknown','no')

In [71]:
data.contact.value_counts()

cellular     26144
telephone    15044
Name: contact, dtype: int64

In [72]:
data.month.value_counts()

may    13769
jul     7174
aug     6178
jun     5318
nov     4101
apr     2632
oct      718
sep      570
mar      546
dec      182
Name: month, dtype: int64

In [73]:
data.day_of_week.value_counts()

thu    8623
mon    8514
wed    8134
tue    8090
fri    7827
Name: day_of_week, dtype: int64

In [74]:
data.job.value_counts()

admin.           10420
blue-collar       9253
technician        6742
services          3966
management        2923
retired           1717
entrepreneur      1456
self-employed     1421
housemaid         1059
unemployed        1014
student            873
unknown            329
SERVICES             3
RETIRED              3
ADMIN.               2
STUDENT              2
MANAGEMENT           1
UNKNOWN              1
BLUE-COLLAR          1
TECHNICIAN           1
HOUSEMAID            1
Name: job, dtype: int64

In [75]:
data['job']=data['job'].replace('unknown','10420')

In [76]:
data.poutcome.value_counts()

nonexistent    35563
failure         4252
success         1373
Name: poutcome, dtype: int64

In [77]:
data_num= data.select_dtypes('int64','float64')
data_num.isna().value_counts()

age    duration  campaign  previous  y    
False  False     False     False     False    41188
dtype: int64

In [78]:
missing_values=data.isna().sum()
missing_values

age               0
job               0
marital           0
education         0
default           0
housing           0
loan              0
contact           0
month             0
day_of_week       0
duration          0
campaign          0
pdays             0
previous          0
poutcome          0
emp_var_rate      0
cons_price_idx    0
cons_conf_idx     0
euribor3m         0
nr_employed       0
y                 0
dtype: int64

Now lets print the cleaned data

In [79]:
data.head()

age          job  marital  ... euribor3m nr_employed  y
0   44  blue-collar  married  ...     4.963      5228.1  0
1   53   technician  married  ...     4.021      5195.8  0
2   28   management   single  ...     0.729      4991.6  1
3   39     services  married  ...     1.405      5099.1  0
4   55      retired  married  ...     0.869      5076.2  1

[5 rows x 21 columns]

# Section 3- Exploring data with Group by¶

Lets dive deep into few features of data and lets see if we can find any valuable information or not. 

We will use groupby funtion to see if features like job, marital , eductaion has any effect on y or not

In [80]:
data[['job','y']].groupby('job').mean()

y
job                    
10420          0.112462
ADMIN.         0.000000
BLUE-COLLAR    0.000000
HOUSEMAID      0.000000
MANAGEMENT     0.000000
RETIRED        0.333333
SERVICES       0.000000
STUDENT        1.000000
TECHNICIAN     0.000000
UNKNOWN        0.000000
admin.         0.129750
blue-collar    0.068951
entrepreneur   0.085165
housemaid      0.100094
management     0.112213
retired        0.252184
self-employed  0.104856
services       0.081442
student        0.312715
technician     0.108276
unemployed     0.142012

In [81]:
data[['marital','y']].groupby('marital').mean()

y
marital           
divorced  0.103209
married   0.101727
single    0.140041

single customer has more probability to opt for term deposit while married and divorced has same variation 

In [82]:
data[['education','y']].groupby('education').mean()

y
education                    
basic-4y             0.083333
basic-6y             0.000000
basic-9y             0.000000
basic.4y             0.102546
basic.6y             0.082240
basic.9y             0.078376
high.school          0.108355
illiterate           0.222222
professional.course  0.113485
university.degree    0.138211

from above analysis we can say that illitrate people have moew probability that they will opt for term deposit but. courses like basic.4y,6y,9y, high school professiona;.course, and university degree have almost same probability to opt for term deposit which is around 10 percent

# section 3 EDA

Lets create a bar graph of all the categorical features with respect to purchase

In [116]:
X = data.iloc[:,[0,20]].values
y = data.iloc[:, 20].values

In [117]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 0)

In [118]:
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

In [119]:
from sklearn.linear_model import LogisticRegression
classifier=LogisticRegression(random_state=0)
classifier.fit(X_train,y_train)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=0, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [120]:
y_pred=classifier.predict(X_test)

In [121]:
y_pred

array([0, 0, 0, ..., 1, 0, 0])

In [122]:
# Importing the dependancies
from sklearn import metrics
# Printing the confusion matrix
# The columns will show the instances predicted for each label,
# and the rows will show the actual number of instances for each label.
print(metrics.confusion_matrix(y_test, y_pred, labels=[0, 1]))
# Printing the precision and recall, among other metrics
print(metrics.classification_report(y_test, y_pred, labels=[0, 1]))

[[9156    0]
 [   0 1141]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00      9156
           1       1.00      1.00      1.00      1141

    accuracy                           1.00     10297
   macro avg       1.00      1.00      1.00     10297
weighted avg       1.00      1.00      1.00     10297

